### PR TITLE
Added new result awaiting and setting ViewModel navigation

### PR DESCRIPTION
**Pull request by entdark**
### :sparkles: What kind of change does this PR introduce? (Bug fix, feature, docs update...)
Adds new `IMvxResultAwaitingViewModel<TResult>` and `IMvxResultSettingViewModel<TResult>` interfaces and corresponding abstract classes to create a result navigation contract between the awaiting and setting the result ViewModels. Under the hood it uses a new `IMvxResultViewModelManager` that registers, unregisters and sets result to registered result awaiting ViewModels.

It's better to use predefined abstract classes `MvxResultAwaitingViewModel<TResult>`, `MvxMultiResultAwaitingViewModel<TResult1, TResult2>`, `MvxResultSettingViewModel<TResult>`, `MvxNavigationResultAwaitingViewModel<TResult>`, `MvxNavigationResultSettingViewModel<TResult>`, and their overloads with parameters. Because when implementing `IMvxResultAwaitingViewModel<TResult>` and `IMvxResultSettingViewModel<TResult>` manually it requires implementing ViewModels registering, saving, restoration, unregistering, and setting the result manually as well.

### :arrow_heading_down: What is the current behavior?
No navigation with awaiting the result.

### :new: What is the new behavior (if this is a feature change)?
Navigation with awaiting the result.

### :boom: Does this PR introduce a breaking change?
No.

### :bug: Recommendations for testing
0. Prepare an Android 13+ device or emulator since we need runtime Notification permissions.
1. Deploy the Playground.Droid app.
2. Start it **without** debugging since the app will be shut down later.
3. Click "SHOW CHILD" button - it will show a child `Fragment` on bottom
4. Minimize the app (not close) by clicking Home or Overview.
5. Go to the Playground.Droid app system settings.
6. Make sure Notifications permissions are granted.
7. Disable Notifications permissions.
8. Go back to the app.
9. The app will be restored with the child `Fragment` on bottom.
10. Click "CLOSE" in the child `Fragment`.
11. Check the logcat (via your preferred IDE): it should display
```
Info (26885) / RootViewModel: [Information] Result Playground.Core.Models.SampleModel from Playground.Core.ViewModels.ChildViewModel
```
12. We received the awaited result after the full app restoration (with shutting down).

You can test this with any dangerous permissions.
Developer option "Don't keep activities" works differently - it does not shut down the app. Though the feature also works in this case.

For other platforms it just works without restoration issues.

### :memo: Links to relevant issues/docs
PRs: #4312 #4652 
Issues: #3342 #3980 #4230 #4262 #4553 #4651 

### :thinking: Checklist before submitting
- [X] All projects build
- [X] Follows style guide lines ([code style guide](https://github.com/MvvmCross/MvvmCross#code-style-guidelines))
- [X] Relevant documentation was updated ([docs style guide](https://www.mvvmcross.com/documentation/contributing/mvvmcross-docs-style-guide))
- [X] Rebased onto current develop

### 🧠 Extra
I am open for discussing changes for the implementation.
Here are some ideas:
- move 3 new navigation extensions to the navigation service directly
- handle default implementations differently through default interface declaration
- rename those long interface and class names to shorter ones

### CODE REVIEW

**Comment by Cheesebaron** [comment]
Thanks will have a look at reviewing this one soon.

**Comment by picosmos** [comment]
@Cheesebaron Can you say when you think you'll be looking at this?

We are currently migrating a project and are considering whether to use the code from this PR or our own implementation from another project. (it's a similar strategy, but we would prefer to use framework features if possible. But unfortunately we can't wait that long anymore.)

**Comment by entdark** [comment]
@picosmos
I think if you hurry then it's better to go with your own.
I am not pleased with the solution in a few places and also probably there will be review fixes.

**Comment by Cheesebaron** [comment]
> @Cheesebaron Can you say when you think you'll be looking at this?
> 
> We are currently migrating a project and are considering whether to use the code from this PR or our own implementation from another project. (it's a similar strategy, but we would prefer to use framework features if possible. But unfortunately we can't wait that long anymore.) Thank you for your great work :)

Sorry, I won't rush a review just because someone needs it. This looks something that would be easy to add to your own presenter override or use an alternative. Even if it gets reviewed and merged in, there is no guarantee there is a release with this in soon either.

I hope you understand :)

**Comment by picosmos** [comment]
> > @Cheesebaron Can you say when you think you'll be looking at this?
> > We are currently migrating a project and are considering whether to use the code from this PR or our own implementation from another project. (it's a similar strategy, but we would prefer to use framework features if possible. But unfortunately we can't wait that long anymore.) Thank you for your great work :)
> 
> Sorry, I won't rush a review just because someone needs it. This looks something that would be easy to add to your own presenter override or use an alternative. Even if it gets reviewed and merged in, there is no guarantee there is a release with this in soon either.
> 
> I hope you understand :)

Sure, that's the right way to do. I think we use our own implementation.

**Comment by amirvenus** [comment]
> > > @Cheesebaron Can you say when you think you'll be looking at this?
> > > We are currently migrating a project and are considering whether to use the code from this PR or our own implementation from another project. (it's a similar strategy, but we would prefer to use framework features if possible. But unfortunately we can't wait that long anymore.) Thank you for your great work :)
> > 
> > 
> > Sorry, I won't rush a review just because someone needs it. This looks something that would be easy to add to your own presenter override or use an alternative. Even if it gets reviewed and merged in, there is no guarantee there is a release with this in soon either.
> > I hope you understand :)
> 
> Sure, that's the right way to do. I think we use our own implementation.

I think one approach would be to add a PR for a new Mvx demo app (similar to the Tip Calculator) using an implementation that wouldn't require changes to the Mvx code.

**Comment by entdark** [comment]
@amirvenus
> I think one approach would be to add a PR for a new Mvx demo app (similar to the Tip Calculator) using an implementation that wouldn't require changes to the Mvx code.

Check #4553 - there are solutions in comments.

File: MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MS-PL license.
+// See the LICENSE file in the project root for more information.
+
+namespace MvvmCross.ViewModels.Result;
+
+public class MvxResultViewModelManager : IMvxResultViewModelManager
+{
+    private readonly Dictionary<Type, HashSet<IMvxViewModel>> _registrations = new();
```
Comment (Cheesebaron) [inline MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs:9]: Are there any implications having strong references to ViewModels in this dict?

File: MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    public void RegisterToResult<TResult>(IMvxResultAwaitingViewModel<TResult> viewModel)
+    {
+        if (!_registrations.TryGetValue(typeof(TResult), out var resultRegistrations))
```
Comment (Cheesebaron) [inline MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs:13]:
```suggestion
        if (!_registrations.TryGetValue(typeof(TResult), out HashSet<IMvxViewModel> resultRegistrations))
```

File: MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            resultRegistrations = new();
+            _registrations[typeof(TResult)] = resultRegistrations;
+        }
+
+        resultRegistrations.Add(viewModel);
+    }
+
+    public bool UnregisterToResult<TResult>(IMvxResultAwaitingViewModel<TResult> viewModel)
+    {
+        if (_registrations.TryGetValue(typeof(TResult), out var resultRegistrations) &&
```
Comment (Cheesebaron) [inline MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs:24]:
```suggestion
        if (_registrations.TryGetValue(typeof(TResult), out HashSet<IMvxViewModel> resultRegistrations) &&
```

File: MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            resultRegistrations.Remove(viewModel))
+        {
+            return true;
+        }
+        return false;
+    }
+
+    public bool IsRegistered<TResult>(IMvxResultAwaitingViewModel<TResult> viewModel)
+    {
+        if (_registrations.TryGetValue(typeof(TResult), out var resultRegistrations) &&
```
Comment (Cheesebaron) [inline MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs:34]:
```suggestion
        if (_registrations.TryGetValue(typeof(TResult), out HashSet<IMvxViewModel> resultRegistrations) &&
```

**Review by Cheesebaron — COMMENTED** [review]
Hey! Sorry for the long wait reviewing the code. Here's a few thoughts about the solution. Let me know what you think.

**Comment by entdark** [comment]
@Cheesebaron hi, I saw the review comments, thanks. I will reply soon.

File: MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs (hunk 1 above)
Comment (entdark) [reply MvvmCross/ViewModels/Result/MvxResultViewModelManager.cs:9]: We need to find which VMs to notify by the result type in `MvxResultViewModelManager.SetResult`. Notice that `MvxResultViewModelManager.SetResult` takes `IMvxResultSettingViewModel` parameter while other methods take `IMvxResultAwaitingViewModel`. `IMvxResultSettingViewModel` doesn't know anything about the awaiter, and the only thing in common for them is `TResult` - that's why we need to store VMs associated to the `TResult` type.

File: MvvmCross/ViewModels/Result/MvxResultSettingViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MS-PL license.
+// See the LICENSE file in the project root for more information.
+
+namespace MvvmCross.ViewModels.Result;
```
[the remark this answers is not in the document]
Comment (entdark) [reply MvvmCross/ViewModels/Result/MvxResultSettingViewModel.cs:5]: Yeah, for convenience. And yes, the default implementation could be moved to the interface and leave `MvxResultSettingViewModel` just inherited from `IMvxResultSettingViewModel` but empty inside.

File: MvvmCross/ViewModels/Result/IMvxResultAwaitingViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+/// public override ViewDestroy(bool viewFinishing = true)
+/// {
+///     base.ViewDestroy();
+///     if (viewFinishing)
+///         this.UnregisterToResult&lt;<typeparamref name="TResult"/>&gt;();
+/// }
+/// </code>
+/// </summary>
+/// <typeparam name="TResult"></typeparam>
+public interface IMvxResultAwaitingViewModel<TResult> : IMvxViewModel
```
[the remark this answers is not in the document]
Comment (entdark) [reply MvvmCross/ViewModels/Result/IMvxResultAwaitingViewModel.cs:29]: That is needed to register VMs in `MvxResultViewModelManager._registrations`. Probably not the best way. I could make non-generic `IMvxResultAwaitingViewModel` that `IMvxResultAwaitingViewModel<TResult>` inherits from and use the same in `MvxResultViewModelManager._registrations` instead of `IMvxViewModel`. But then it's kind of incorrect since `IMvxResultAwaitingViewModel` supposes to have a result to await. End users may be confused which to use.
On the other hand I could use `object` in `MvxResultViewModelManager._registrations` instead of `IMvxViewModel`

[the remark this answers is not in the document]
Comment (Cheesebaron) [reply MvvmCross/ViewModels/Result/IMvxResultAwaitingViewModel.cs:29]: Maybe instead of using `IMvxViewModel` as a common base, then use a new interface. Then you have something for `MvxResultViewModelManager` to look for, while not depending on `IMvxViewModel`.

File: MvvmCross/ViewModels/Result/MvxResultViewModelExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    }
+
+    public static void SaveRegisterToResult<TResult>(this IMvxResultAwaitingViewModel<TResult> viewModel, IMvxBundle savedStateBundle)
+    {
+        if (Mvx.IoCProvider?.TryResolve<IMvxResultViewModelManager>(out var manager) == true &&
+            manager.IsRegistered(viewModel) &&
+            savedStateBundle?.Data is { } data)
+        {
+            data[BundleRegisterKey] = true.ToString();
```
Comment (Cheesebaron) [inline MvvmCross/ViewModels/Result/MvxResultViewModelExtensions.cs:27]: Could this be a const instead?

**Comment by Cheesebaron** [comment]
I think I am OK with this implementation. One last thing, could you add some docs for this with examples of usage. That would help anyone with adopting this.

File: MvvmCross/ViewModels/Result/MvxResultViewModelExtensions.cs (hunk 7 above)
Comment (entdark) [reply MvvmCross/ViewModels/Result/MvxResultViewModelExtensions.cs:27]: if you mean `BundleRegisterKey` then it's already `const` defined above.

Comment (Cheesebaron) [reply MvvmCross/ViewModels/Result/MvxResultViewModelExtensions.cs:27]: No the `true.ToString()` it will always result in `"true"` :)

**Comment by follesoe** [comment]
It would be awesome if @entdark could push it to the last mile. This would help migrate code to MvvmCross 9 🙌 (as far as I can see it's only some documentation missing). Highly appreciate the effort done so far.

**Comment by Cheesebaron** [comment]
I've gone through the code and upon playing with it, I might change it some to address some of the issues around having some of the logic in extension methods which isn't possible to override the behavior off. So I think I will merge this and finish it myself. This is not to discount the work @entdark has done. I really appreciate the work done here, I just need to ensure everyone is successful using this feature for the future.

**Comment by entdark** [comment]
I still plan to document, and remember about the issue, I just did not have free time lately. I was waiting for the PR review for months so it would be nice if you waited for my documentations as well. Unless you want to do it your way. @Cheesebaron

**Comment by Cheesebaron** [comment]
So, I've gone ahead and removed the hidden dependencies in the Extensions.

Another thing I experience, is in the Android sample, the result never gets set from the child model, do you have any idea why this is happening @entdark?

**Comment by entdark** [comment]
@Cheesebaron
And I don't quite understand your changes and necessity of explicitly passing `IMvxResultViewModelManager`, but it's up to you ofc.
Maybe you opened the wrong VM?
And have you tested w/o your changes?

**Comment by Cheesebaron** [comment]
The hidden dependencies in extension methods are not preferred because it hides dependencies. But also does not allow for overwriting behavior like you can most places in MvvmCross.

As for opening wrong VM, I don't think so. Opened the only instance of ChildVM there is in RootVM. Nothing is returned. Made sure OnDestroy is called on the Fragment

**Comment by entdark** [comment]
Are you saying that the result is not set after the restoration or within just first app life w/o closing the app at all?

**Comment by Cheesebaron** [comment]
> Are you saying that the result is not set after the restoration or within just first app life w/o closing the app at all?

In this case, I am simply opening the child VM using the first button in the list. Then pressing close button. Nothing is returned in `ResultSet`.


https://github.com/user-attachments/assets/b334b770-ff5c-4ef5-9706-77f23be755ef

**Comment by entdark** [comment]
@Cheesebaron I will summarize what I wrote on Discord about this.
I found the issue of failing setting the result:
When `RootViewModel` navigates to `ChildViewModel` it registers to result when navigation is completed: https://github.com/entdark/MvvmCross/blob/new_result_navigation/MvvmCross/Navigation/MvxNavigationExtensions.cs#L88-L91
But navigation gets completed when `Initialize` task get completed and it's delayed by 8,5 seconds: https://github.com/entdark/MvvmCross/blob/new_result_navigation/Projects/Playground/Playground.Core/ViewModels/Navigation/ChildViewModel.cs#L60
So, if you wait 8,5 seconds and then close the `ChildViewModel` then result will be set correctly.
Currently it fails to set the result when `ChildViewModel` gets closed fast.

After some thinking I got that it's kind of weird that navigation depends on `Initialize` completion, especially if it fails at some point but visually navigation could be done, but `navigated` will be `false` at that point.
I think navigation should be just navigation and `Initialize` should not block it and go just afterward failing or not
So, to summarize it's better to change the navigation logic and no need to change anything in the `NavigateRegisteringToResult`.

**Comment by Cheesebaron** [comment]
Added a separate sample for the navigation with result to not pollute existing ones.

**Comment by entdark** [comment]
That is still an issue with heavy loading in `Initialize`

**Comment by Cheesebaron** [comment]
> That is still an issue with heavy loading in `Initialize` in real projects.

I think if it is well enough documented, most people will have a pretty good time with this feature. So I think that is the last thing missing.

**Comment by entdark** [comment]
But why would a failed `Initialize` fail navigation task? If visually we already saw the page appeared, it's just the logic behind that does not mark it as successful navigation. I think it can lead into more new unnecessary issues/discussions in the future.

**Comment by Cheesebaron** [comment]
> But why would a failed `Initialize` fail navigation task? If visually we already saw the page appeared, it's just the logic behind that does not mark it as successful navigation. I think it can lead into more new unnecessary issues/discussions regarding result navigation in the future.

I am not 100% certain why it was done like this originally. My guess is to not trigger other lifecycle methods before Initialize task is fully finished.

Changing it to fire and forget would also be a breaking change, so I am inclined not to touch it for now. I'd rather do it in another PR.

**Comment by entdark** [comment]
> I am inclined not to touch it for now

👌
Then I am fine with solving problems after their appearance if any.

If code part is settled then all I need is to document, right? I got some free time now.

**Comment by Cheesebaron** [comment]
> > I am inclined not to touch it for now
> 
> 👌 Then I am fine with solving problems after their appearance if any.
> 
> If code part is settled then all I need is to document, right? I got some free time now.

Yeah, I don't think I am going to make more changes to this. Please go ahead and add some documentation for this feature.

**Comment by entdark** [comment]
There you go.
